### ISSUE TITLE
How to package a completely external NER model with spacy for use in prodigy

### ISSUE THREAD

**Issue by hannahlindsley** [explosion/spaCy#2676]
I'm unclear on how I can take an existing, external-to-spacy NER model (a crf) and package it for use in prodigy. I've been able to do it for a custom tokenizer by extending the spacy Tokenizer class, but I think I'm barking up all kinds of wrong trees when I try to do it for NER. Am I missing this in the doc? All I've been able to ascertain so far is how to _retrain_ an existing spacy model, which is not what I'm looking to do. Thanks!

## Which page or section is this issue related to?

https://spacy.io/usage/training#section-ner
https://spacy.io/usage/training#saving-loading
https://spacy.io/api/language#to_disk

**Comment by ines** [comment]
Hi! This is definitely something that's currently under-documented, also because it used to be much more difficult. But I've been meaning to write a blog post on this topic that shows and end-to-end example of packaging a fully custom component with a model, so the timing of this question is actually really good!

One quick question before I write more: Could you share an example of the usage of your custom model? How does it currently integrate with spaCy and how do you call it from Python? Is it a pipeline component?

**Comment by hannahlindsley** [comment]
I actually just figured it out! (I loooove your work, btw, I'm constantly learning new ways to use it.)

Here's how I did it:
`
prefix_re = regex.compile(r"^" + token_regex)
suffix_re = regex.compile(token_regex + r"$")
infix_re = regex.compile(token_regex)
# this is kludgy as heeeelllll - must be a better way to say "None"
simple_re = regex.compile(r'''____&&&976576345646u567546u''')

class CustomTokenizer(Tokenizer):

    def __init__(self, nlp):
        super().__init__(nlp.vocab,
                         rules={},
                         prefix_search=prefix_re.search,
                         suffix_search=suffix_re.search,
                         infix_finditer=infix_re.finditer,
                         token_match=simple_re.match)

    def __call__(self, string, *args, **kwargs):
        return super().__call__(xml.unescape(html.unescape(string)))

class CustomEntityRecognizer:

    # don't actually have any idea what of this is necessary
    def __init__(self, nlp):
        self.vocab = nlp.vocab

    # this is the important stuff! I define a function from spacy_doc to label span, and call it here.
    def __call__(self, spacy_doc, *args, **kwargs):
        return predict_single(spacy_doc)

nlp = spacy.load('en_core_web_sm', disable=['ner', 'parser', 'tagger'])
nlp.tokenizer = CustomTokenizer(nlp)
nlp.factories['CustomEntityRecognizer'] = CustomEntityRecognizer
nlp.create_pipe('CustomEntityRecognizer')
nlp.add_pipe(CustomEntityRecognizer, last=True)

model_directory = '/path/to/custom_entity_recognizer'
nlp.to_disk(model_directory)

nlp = spacy.load(model_directory)
print(nlp("this is an entity for detection hello@gmail.com plus some html &gt;"))
`

**Comment by hannahlindsley** [comment]
I actually just figured it out! I am definitely not clear on which of these things I actually need, or why I'm doing some of them... but it's working...
Love your work, by the way, I'm constantly learning new applications for it. Thanks so much!

`
	prefix_re = regex.compile(r"^" + token_regex)
	suffix_re = regex.compile(token_regex + r"$")
	infix_re = regex.compile(token_regex)
	# this is kludgy as heeeelllll - must be a better way to say "None"
	simple_re = regex.compile(r'''____&&&976576345646u567546u''')

	class CustomTokenizer(Tokenizer):

	    def __init__(self, nlp):
	        super().__init__(nlp.vocab,
	                         rules={},
	                         prefix_search=prefix_re.search,
	                         suffix_search=suffix_re.search,
	                         infix_finditer=infix_re.finditer,
	                         token_match=simple_re.match)

	    def __call__(self, string, *args, **kwargs):
	        return super().__call__(xml.unescape(html.unescape(string)))

	class CustomEntityRecognizer:

	    # don't actually have any idea what of this is necessary
	    def __init__(self, nlp):
	        self.vocab = nlp.vocab

	    # this is the important stuff! I define a function from spacy_doc to label span, and call it here.
	    def __call__(self, spacy_doc, *args, **kwargs):
	        return predict_single(spacy_doc)

	nlp = spacy.load('en_core_web_sm', disable=['ner', 'parser', 'tagger'])
	nlp.tokenizer = CustomTokenizer(nlp)
	nlp.factories['CustomEntityRecognizer'] = CustomEntityRecognizer
	nlp.create_pipe('CustomEntityRecognizer')
	nlp.add_pipe(CustomEntityRecognizer, last=True)

	model_directory = '/path/to/custom_entity_recognizer'
	nlp.to_disk(model_directory)

	nlp = spacy.load(model_directory)
	print(nlp("this is an entity for detection hello@gmail.com plus some html &gt;"))
`

**Comment by hannahlindsley** [comment]
I actually just figured it out! I am definitely not clear on which of these things I actually need, or why I'm doing some of them... but it's working...
Love your work, by the way, I'm constantly learning new applications for it. Thanks so much!

	prefix_re = regex.compile(r"^" + token_regex)
	suffix_re = regex.compile(token_regex + r"$")
	infix_re = regex.compile(token_regex)
	# this is kludgy as heeeelllll - must be a better way to say "None"
	simple_re = regex.compile(r'''____&&&976576345646u567546u''')

	class CustomTokenizer(Tokenizer):

	    def __init__(self, nlp):
	        super().__init__(nlp.vocab,
	                         rules={},
	                         prefix_search=prefix_re.search,
	                         suffix_search=suffix_re.search,
	                         infix_finditer=infix_re.finditer,
	                         token_match=simple_re.match)

	    def __call__(self, string, *args, **kwargs):
	        return super().__call__(xml.unescape(html.unescape(string)))

	class CustomEntityRecognizer:

	    def __init__(self, nlp):
	        self.vocab = nlp.vocab

	    # this is the important stuff! I define a function from spacy_doc to list of entities, and call it here. See sample entity at bottom 
	    def __call__(self, spacy_doc, *args, **kwargs):
	        return predict_single(spacy_doc)

	nlp = spacy.load('en_core_web_sm', disable=['ner', 'parser', 'tagger'])
	nlp.tokenizer = CustomTokenizer(nlp)
	nlp.factories['CustomEntityRecognizer'] = CustomEntityRecognizer
	nlp.create_pipe('CustomEntityRecognizer')
	nlp.add_pipe(CustomEntityRecognizer, last=True)

	model_directory = '/path/to/custom_entity_recognizer'
	nlp.to_disk(model_directory)

	nlp = spacy.load(model_directory)
	print(nlp("this is an entity for detection hello@gmail.com plus some html &gt;"))

This returns (per my predict_single function)
    [{'label': 'EMAIL', 'begin': 32, 'end': 67, 'token_begin': 6, 'token_end': 16}]

**Comment by hannahlindsley** [comment]
I'm currently using prodigy to label all of our raw training data. I can see the results of the model through prodigy as well, because I've written a custom recipe that pulls in the model - that was not ideal, as the prodigy setup repo doesn't really need to have my custom model code in it; I'd rather specify which model to use in the recipe from the command line (this also lets me try out new versions of the model all the time without changing any code). In order to do so, it seemed from the doc that the model would need to be spacy-ified (packaged as a spacy model), which is hopefully what some version of the code above will give me.

**Comment by ines** [comment]
Yes, that looks good so far – and your analysis is definitely correct. All custom components and their data need to be available in the model package. In the case of your custom tokenizer, this is easy, because spaCy knows how to serialize the tokenizer rules as JSON. But for any of your custom components, it doesn't (and it won't just dump all your code and eval it by default). So the best way to add custom code and logic to a model is to wrap it as a Python package.

If you run the [`spacy package`](https://spacy.io/api/cli#package) command on a model directory you saved out, it will create all files you need for a Python package. This includes the [`__init__.py`](https://github.com/explosion/spacy-models/blob/master/template/xx_model_name/__init__.py), which mostly consists of a `load()` function. When you call `spacy.load` on an installed model package, this is the function that will be executed to load it. So running `spacy.load('en_core_web_sm')` is equivalent to:

```python
import en_core_web_sm
nlp = en_core_web_sm.load()
```

So the quick and dirty solution would be to add all your custom code to the model's `__init__.py` and add the `CustomEntityRecognizer` to the global factories, so spaCy knows how to initialise the component when it loads your model:

```python
from spacy.language import Language

# add your custom NER to the global factories
Language.factories['CustomEntityRecognizer'] = CustomEntityRecognizer
```

*Everything* needs to be available from within the package, though – including your `predict_single` function and everything else it needs. You can also specify additional dependencies in the `setup.py`, add multiple files for multiple modules etc.

Once you're done editing your Python package, you can run the following command from within the directory:

```bash
python setup.py sdist
```

This will build the package and add a `.tar.gz` archive to a directory `/dist`. The archive should include everything in the package, and can be installed via `pip install`. To make spaCy load the model via its `__init__.py`, it needs to be *installed as a package* – so you need to run `pip install your_model.tar.gz` within your Prodigy environment. You'll then be able to load it by its package name, e.g. `your_model`.

### Making things more elegant

Of course, dumping all your code into the package isn't that elegant, so there are various ways to make the whole thing nicer and more maintainable. But it really depends on how far you want to take it and what your requirements are.

#### Adding (serialization) methods to your custom component

If your custom model needs to load data, you can implement all of this directly via your custom component, by adding `to_disk`, `from_disk`, `to_bytes` and `from_bytes` methods. When you call `nlp.from_disk`, spaCy essentially does the following – it iterates over the pipeline components, checks whether they have a `from_disk` method implemented, and if so, calls it:

```python
for pipe_name, proc in nlp.pipeline:
   if hasattr(proc, 'from_disk'):
       proc.from_disk(model_path / pipe_name)
```

So if your custom component is called `"custom_ner"`, spaCy will call its `from_disk` method and pass it the model path plus the component name, so `/path/to/model/custom_ner`. The component can then load stuff from there – e.g. one binary file, several files, whatever it needs.

```python
class CustomEntityRecognizer:
    name = 'custom_ner'

    def __init__(self, nlp):
        self.vocab = nlp.vocab
        self.some_data = None

    def __call__(self, spacy_doc, *args, **kwargs):
        return predict_single(spacy_doc)

    def from_disk(self, path, *kwargs):
        # do some stuff here and load all the data the component needs
        data_path = path / 'some_data.json'
        with data_path.open() as f:
            self.some_data = json.load(f)
```

You can find an example of a component that implements all of these methods in #2513 (a component that takes a list of patterns and labels the matches as entities).

In theory, your component can also implement all other methods of the [`Pipe`](https://spacy.io/api/pipe) base class – for example `update` and `begin_training` to make it trainable.

#### Shipping the component as a separate package

If you want to reuse the pipeline component and code with different data, you can also make it its own package and then have your model package require your component package.

#### Shipping model-specific Prodigy recipes with the model

Maybe that's overkill, but Prodigy supports providing recipes via Python entry points – i.e. modules specified in the `setup.py` of another package. So if you're using a Prodigy recipe that's super specific to your model, you could ship a `prodigy_recipes.py` *with the model package*. If that package is installed in the same directory as Prodigy, Prodigy will find the recipe automatically via its name. You can find more details and examples of this in the "Entry points" section of the `PRODIGY_README.html`.

### TL;DR

I'd suggest the following workflow:

1. Save out the `en_core_web_sm` model with your custom tokenizer only.
2. Run `spacy package` on the saved directory, edit the meta accordingly etc.
3. Open the package directory and edit the model's `__init__.py`. Include your custom component, custom NER model etc. and add add an entry to the global factories.
4. Run `python setup.py sdist` within your package directory to build the package.
5. Install the `.tar.gz` model archive created in `/dist` in your Prodigy directory. You should now be able to load it by name, and your code should be executed on load.
6. Optional: make things more elegant, if you like.
